# ExploreSCIS 2021
# Basic Geospatial Visualisation Workshop by SMU Anthill Society



In this notebook, we will be using dataset of hawker centres retrieved from data.gov.sg (https://data.gov.sg/dataset/dates-of-hawker-centres-closure) to do an exploratory data analysis with Pandas and geospatial analysis with Folium.

#Exploratory Data Analysis
Exploratory Data Analysis (EDA) is an approach to analyse datasets to summarise their main characteristics, usually early in an analytical process. Other than giving us a better understanding of the data, it also helps us to detect mistakes and check assumptions. We can also see relationships among the variables in the dataset.

In this workshop, we will be doing a simple form of EDA with the Pandas library.

#Introduction to Pandas
Pandas is a Python library mainly used for data analysis. It is a fast, powerful, flexible and easy to use open source data analysis and manipulation tool, built on top of the Python programming language. 

## 1. Importing the Pandas Library 

To use Pandas, we will need to install and import the pandas library. The Pandas library is already available in Google Colaboratory hence, we just need to import it. While it is not necessary to import pandas as pd, it is a convention to do so (and you can type less).

In [ ]:
import pandas as pd

## 2. Reading the data into a Pandas DataFrame

###2a. Reading in the data

We will load our data into the Pandas DataFrame, which is a 2-dimensional labeled data structure with rows and columns. It is a good visual representation of data as it is like a table.

To load the data we have, we will use the **`.read_csv()`** function since our data is in a csv file. 
<br> Read more about the function here: https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html


In [ ]:
url = 'https://raw.githubusercontent.com/wnyeo/ExploreSCIS_data/main/hawker_final.csv' # link to the cleaned dataset

hawker_df = pd.read_csv(url)  # df represents DataFrame

In [ ]:
hawker_df # showing the data

name  ...                         google_3d_view
0                                Adam Road Food Centre  ...  https://goo.gl/maps/ZMLJeP8STKpDP9so9
1    Aljunied Ave 2 Blk 117 (Blk 117 Aljunied Marke...  ...  https://goo.gl/maps/XVjxu6hgpcwt15JC8
2     Amoy Street Food Centre (Telok Ayer Food Centre)  ...  https://goo.gl/maps/RiX319zQXRFeHWPE6
3    Ang Mo Kio Ave 1 Blk 226D (Kebun Baru Market a...  ...  https://goo.gl/maps/CAJGbiEU4o46xfbR8
4           Ang Mo Kio Ave 1 Blk 341 (Teck Ghee Court)  ...  https://goo.gl/maps/quL2H4mojGZfZsHA6
..                                                 ...  ...                                    ...
109  Whampoa Drive Blk 90 (Whampoa Drive Makan Plac...  ...  https://goo.gl/maps/PQXB1ifE7PrnMbAf7
110  Whampoa Drive Blk 91/92 (Whampoa Drive Makan P...  ...  https://goo.gl/maps/vERiFeXoGi8jDfjJ8
111                          Yishun Park Hawker Centre  ...  https://goo.gl/maps/jqWXYVF17BiAz7bP9
112  Yishun Ring Road Blk 104/105 (Chong Pang Marke...  ...  https://goo.gl/maps/dPjV2SSaX9nEXV1u8
113                         Zion Riverside Food Centre  ...  https://goo.gl/maps/8kwxWX9gy3J3Mna58

[114 rows x 10 columns]

### 2b. `head()` Function

To show only the first n rows of data, we can make use of the **`.head()`** function. By default, `df.head()` will return the first 5 rows of the DataFrame. If you want more/less number of rows, you can specify an integer.

In [ ]:
hawker_df.head(3)

name  ...                         google_3d_view
0                              Adam Road Food Centre  ...  https://goo.gl/maps/ZMLJeP8STKpDP9so9
1  Aljunied Ave 2 Blk 117 (Blk 117 Aljunied Marke...  ...  https://goo.gl/maps/XVjxu6hgpcwt15JC8
2   Amoy Street Food Centre (Telok Ayer Food Centre)  ...  https://goo.gl/maps/RiX319zQXRFeHWPE6

[3 rows x 10 columns]

### 2c. `tail()` Function [Mini Exercise!][2 mins]

To show the last n rows of data, we can make use of the **`.tail()`** function. Similarly, by default, `df.tail()` will return the last 5 rows of the DataFrame. If you want more/less number of rows, you can specify an integer.
<br>
<br>**Exercise: Show the last 7 rows of hawker_df** 

In [ ]:
# Enter your code here

####Solution
Click below for the solution.

In [ ]:
hawker_df.tail(7)

name  ...                         google_3d_view
107       West Coast Drive Blk 502 (Ayer Rajah Market)  ...  https://goo.gl/maps/dWso9emez66FrUyJ8
108  West Coast Drive Blk 503 (Ayer Rajah Food Centre)  ...  https://goo.gl/maps/HzZpoN9MxHR4ie9v6
109  Whampoa Drive Blk 90 (Whampoa Drive Makan Plac...  ...  https://goo.gl/maps/PQXB1ifE7PrnMbAf7
110  Whampoa Drive Blk 91/92 (Whampoa Drive Makan P...  ...  https://goo.gl/maps/vERiFeXoGi8jDfjJ8
111                          Yishun Park Hawker Centre  ...  https://goo.gl/maps/jqWXYVF17BiAz7bP9
112  Yishun Ring Road Blk 104/105 (Chong Pang Marke...  ...  https://goo.gl/maps/dPjV2SSaX9nEXV1u8
113                         Zion Riverside Food Centre  ...  https://goo.gl/maps/8kwxWX9gy3J3Mna58

[7 rows x 10 columns]

###2d. Checking the size of data

We can check the size of the data using the **`.shape`** attribute

In [ ]:
hawker_df.shape

(114, 10)

We can see that there are 114 rows and 10 columns in our dataset

## 3. Accessing data in the Dataframe

###3a. Accessing rows and columns
`loc[:]` helps us to access a group of rows and columns in a dataset. For instance, if we only want the first 2 rows and the first columns of a dataset, we can use `loc[:]` to do so. 

In [ ]:
hawker_df.loc[0:1, ['name']]

name
0                              Adam Road Food Centre
1  Aljunied Ave 2 Blk 117 (Blk 117 Aljunied Marke...

### 3b. Accessing individual columns/columns only

We can also access columns directly using square brackets

In [ ]:
hawker_df['name']

0                                  Adam Road Food Centre
1      Aljunied Ave 2 Blk 117 (Blk 117 Aljunied Marke...
2       Amoy Street Food Centre (Telok Ayer Food Centre)
3      Ang Mo Kio Ave 1 Blk 226D (Kebun Baru Market a...
4             Ang Mo Kio Ave 1 Blk 341 (Teck Ghee Court)
                             ...                        
109    Whampoa Drive Blk 90 (Whampoa Drive Makan Plac...
110    Whampoa Drive Blk 91/92 (Whampoa Drive Makan P...
111                            Yishun Park Hawker Centre
112    Yishun Ring Road Blk 104/105 (Chong Pang Marke...
113                           Zion Riverside Food Centre
Name: name, Length: 114, dtype: object

In [ ]:
hawker_df[['name', 'address_myenv']]

name                                      address_myenv
0                                Adam Road Food Centre                     2, Adam Road, Singapore 289876
1    Aljunied Ave 2 Blk 117 (Blk 117 Aljunied Marke...          Blk 117, Aljunied Ave 2, Singapore 380117
2     Amoy Street Food Centre (Telok Ayer Food Centre)  National Development Building, Annex B, Telok ...
3    Ang Mo Kio Ave 1 Blk 226D (Kebun Baru Market a...       Blk 226D, Ang Mo Kio Ave 1, Singapore 564226
4           Ang Mo Kio Ave 1 Blk 341 (Teck Ghee Court)        Blk 341, Ang Mo Kio Ave 1, Singapore 560341
..                                                 ...                                                ...
109  Whampoa Drive Blk 90 (Whampoa Drive Makan Plac...            Blk 90, Whampoa Drive, Singapore 320090
110  Whampoa Drive Blk 91/92 (Whampoa Drive Makan P...  Blk 91/92, Whampoa Drive, Singapore 320091/320092
111                          Yishun Park Hawker Centre         Blk 51, Yishun Avenue 11, Singapore 768867
112  Yishun Ring Road Blk 104/105 (Chong Pang Marke...  Blk 104/105, Yishun Ring Road, Singapore 76010...
113                         Zion Riverside Food Centre                    70, Zion Road, Singapore 247792

[114 rows x 2 columns]

##4. Understanding the descriptive statistics of the data

Descriptive statistics summarise the information of the characteristics and distribution of values in the dataset. It allows us to learn about the central tendency and variability of the data. Examples include the mean, median, mode, standard deviation, etc.

### 4a. `describe()` function
We can understand the descriptive statistics of the data using the 
**`.describe()`** function

In [ ]:
hawker_df.describe()

latitude_hc  longitude_hc  no_of_market_stalls  no_of_food_stalls
count   114.000000    114.000000           114.000000         114.000000
mean      1.325880    103.844182            72.745614          56.105263
std       0.037271      0.056990            72.218549          32.928823
min       1.272660    103.697374             0.000000           0.000000
25%       1.302355    103.808046             0.000000          36.000000
50%       1.320323    103.845428            69.500000          50.000000
75%       1.344818    103.883722           107.250000          72.000000
max       1.443990    103.988182           477.000000         226.000000

### 4b. `count()` function [Mini Exercise!][2 min]
We can find the number of values for each column using the **`.count()`** function
<br>
<br>**Exercise: Find the count of values for each column in the dataframe** 

In [ ]:
# Enter your code here

####Solution
Click below for the solution.

In [ ]:
hawker_df.count()

name                   114
latitude_hc            114
longitude_hc           114
photourl               114
address_myenv          114
no_of_market_stalls    114
no_of_food_stalls      114
description_myenv      114
status                 114
google_3d_view         114
dtype: int64

Now that we have a basic idea of the dataset that we are working with, we can start to plot the locations of Hawker Centres onto the map

# Introduction to Foilum
Folium is a data visualization library in Python that was built primarily to help people visualize geospatial data. With Folium, one can create a map of any location in the world if its latitude and longitude values are known.

## 5. Importing the Folium library
Similarly to use Folium, we will need to install and import the Folium library. The Folium library is already available in Google Colaboratory hence, we just need to import it. 

In [ ]:
import folium

##6. Initialising a map


### 6a. Map() function
We can generate a base map using the folium library using the **`Map()`** function. 

In [ ]:
map = folium.Map(location=[1.3437459, 103.8240449], # coordinates (latitude and longitude) of the map, these are the coordinates for Singapore
                 zoom_start=12,                     # initial zoom level for the map
                 control_scale=True)                # whether to add a control scale on the map.

### 6b. Displaying the map
Display the map using the **`display()`** function

In [ ]:
display(map)

## 7. Adding the locations onto the map

### 7a. Retrieving the relevant rows for the geospatial visualisation

In [ ]:
hc_locations = hawker_df[["latitude_hc", "longitude_hc", 'name']]

In [ ]:
hc_locations

latitude_hc  ...                                               name
0       1.324132  ...                              Adam Road Food Centre
1       1.320710  ...  Aljunied Ave 2 Blk 117 (Blk 117 Aljunied Marke...
2       1.279129  ...   Amoy Street Food Centre (Telok Ayer Food Centre)
3       1.366950  ...  Ang Mo Kio Ave 1 Blk 226D (Kebun Baru Market a...
4       1.364170  ...         Ang Mo Kio Ave 1 Blk 341 (Teck Ghee Court)
..           ...  ...                                                ...
109     1.323140  ...  Whampoa Drive Blk 90 (Whampoa Drive Makan Plac...
110     1.323490  ...  Whampoa Drive Blk 91/92 (Whampoa Drive Makan P...
111     1.424911  ...                          Yishun Park Hawker Centre
112     1.431658  ...  Yishun Ring Road Blk 104/105 (Chong Pang Marke...
113     1.292340  ...                         Zion Riverside Food Centre

[114 rows x 3 columns]

### 7b. Adding a location onto the base map

Using **`Marker()`**, we can add a location onto the base map that we have generated, by providing the coordinates. There is a parameter named popup where we can assign a text to. The text will be shown when we click on the marker.

In [ ]:
folium.Marker([1.324132, 103.814163], # latitude and longitude
              popup='Adam Road Food Centre' # name of hawker centre, what the popup will display when the marker is clicked
              ).add_to(map) # add it to base map that was initialised

In [ ]:
display(map)

### 7c. Adding multiple locations to the base map with the aid of a for loop
However, it is quite a hassle to add each and every location onto the map individually. But we can speed up the process with the help of a for loop.

In [ ]:
for index, location_info in hc_locations.iterrows(): # iterrows is used to iterate through the rows of the dataframe
    folium.Marker([location_info["latitude_hc"], location_info["longitude_hc"]], # latitude and longitude
                  popup=location_info["name"]  # name of hawker centre
                  ).add_to(map) 

For loops are used for iterating over iterable objects, such as the rows of the DataFrame.




**For a better idea of what `iterrows()` does.** <br> `DataFrame.iterrows` is a generator which yields both the index and row (as a Series)

In [ ]:
for index, location_info in hc_locations.iterrows(): 
  print(index)
  print()
  print(location_info)
  break

0

latitude_hc                   1.32413
longitude_hc                  103.814
name            Adam Road Food Centre
Name: 0, dtype: object


### 7d. Display the map [Mini Exercise!]
**Exercise: Display the map that has the location added to it**

In [ ]:
# Enter your code here

####Solution
Click below for the solution.

In [ ]:
display(map)

#Extra Exercises!

## Plotting a heatmap

To understand the density of the hawker centres spatially, we can use the folium heat map to plot the listings. There are 2 broad steps to creating a heatmap:

i) Creating a folium base map of the location (which we have learnt above)<br>
ii) Adding the heatmap layer

In [ ]:
import folium
from folium.plugins import HeatMap

In [ ]:
lat_long = []
for index, location_info in hc_locations.iterrows():
  lat_long.append([location_info['latitude_hc'], location_info['longitude_hc']])

map_folium = folium.Map([1.35255,103.82580], zoom_start=11, control_scale=True)

HeatMap(lat_long, radius = 8, gradient = {0.2:'blue', 0.3:'purple', 0.5:'orange', 1.0:'red'}).add_to(map_folium)
display(map_folium)

## Geospatial visualisation with other datasets
Here are some other datasets that you can try to visualise
<br> https://raw.githubusercontent.com/wnyeo/ExploreSCIS_data/main/hdb_final.csv
<br> https://raw.githubusercontent.com/wnyeo/ExploreSCIS_data/main/mrt_lrt_final.csv
<br> You can follow the steps outlined above

To challenge yourself further, you can also plot the distribution of MRTs and LRTs and HDBs together to see how accessible our country is :)